File: EPI/12.1/swift/12.1.swift
public func firstIndexOf<T: Comparable>(element: T, inSortedArray array: [T]) -> Int {
    var low = 0
    var high = array.count - 1
    var index = -1

    while low <= high {
        let mid = low + (high - low) / 2
        let midElement = array[mid]

        if midElement < element {
            low = mid + 1
        } else if midElement > element {
            high = mid - 1
        } else {
            index = mid
            high = mid - 1
        }
    }

    return index
}

File: EPI/12.1/typescript/12.1.ts
export default function firstIndexOf<T>(array: T[], element: T): number {
    let left = 0;
    let right = array.length - 1;

    while (right >= left) {
        const mid = Math.floor(left + (right - left) / 2);
        if (array[mid] >= element) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    const index = right + 1;
    return index >= 0 && index < array.length && array[index] === element ? index : -1;
}
